### PR TITLE
GC Tombstone: Add option for apps to pass to disable GC enforcement for old files

### PR DESCRIPTION
**Pull request by markfields**
## Description

In discussions with our internal partner teams, as they consider rolling out GC via the Tombstone feature, they want a way to cut off old documents from exposure to failure due to GC (Tombstones failing to load, or GC Sweep deleting objects).  This is because some GC-impacting bugs in partner code were forward-fixed only, without leaving repair code in place - so documents created before such a bug was fixed would be subject to being corrupted by GC (e.g. due to invalid internal document structure), if they were opened after GC was enabled.

This PR adds an otherwise undocumented ContainerRuntime Option `gcEnforcementMinCreateContainerRuntimeVersion`, which is to be compared with the value of the persisted metadata property `createContainerRuntimeVersion` to determine whether a document was created recently enough to be subjected to enforcement of GC.

Some clarifications:

* If `gcEnforcementMinCreateContainerRuntimeVersion` is omitted (the default case since it's undocumented) or malformed, GC enforcement is determined based on the existing configuration parameters.  Only if it's present and strictly `>` the file's `createContainerRuntimeVersion` will GC enforcement be disabled.
* If `createContainerRuntimeVersion` is missing from a file, it's considered older than any valid value of `gcEnforcementMinCreateContainerRuntimeVersion` such that if the option is set, GC enforcement will be disabled.

## Reviewer Guidance

There are a few todos (`//*` comments) but the basic structure is there.

I'd love special attention given to the gcHelper tests that cover the version comparison stuff.  What other test cases should I put in there?

### CODE REVIEW

File: packages/runtime/container-runtime/src/test/garbageCollection.spec.ts [hunk 1]
```diff
@@ -230,6 +231,28 @@ describe("Garbage Collection Tests", () => {
 				assert(gc.gcEnabled, "gcEnabled incorrect");
 				assert.equal(gc.latestSummaryGCVersion, 1, "latestSummaryGCVersion incorrect");
 			});
+			it("gcEnforcementMinCreateContainerRuntimeVersion later than persisted value, sweepEnabled false", () => {
```
Comment (markfields) [inline packages/runtime/container-runtime/src/test/garbageCollection.spec.ts:234]: This illustrates how GC enforcement would be disabled for an old document.  It's using the `sweepEnabled` setting but would also apply to tombstone failure.

File: packages/test/test-end-to-end-tests/src/test/gc/gcTombstoneDataStores.spec.ts [hunk 2]
```diff
@@ -602,6 +623,51 @@ describeNoCompat("GC data store tombstone tests", (getTestObjectProvider) => {
...
+				// The datastore should be tombstoned now
+				const { summaryVersion } = await summarize(summarizer);
+				const container = await loadContainer(summaryVersion);
+				await sendOpToUpdateSummaryTimestampToNow(container);
+
+				// This request succeeds even though the datastore is tombstoned, on account of the later gcEnforcementMinCreateContainerRuntimeVersion passed in
+				const tombstoneSuccessResponse = await containerRuntime_resolveHandle(container, {
+					url: unreferencedId,
+				});
+				assert.equal(
```
Comment (agarwal-navin) [inline packages/test/test-end-to-end-tests/src/test/gc/gcTombstoneDataStores.spec.ts:658]: The tombstone errors will still be logged right?

File: packages/runtime/container-runtime/src/test/garbageCollection.spec.ts [hunk 3]
```diff
@@ -230,6 +231,28 @@ describe("Garbage Collection Tests", () => {
...
+				gc = createGcWithPrivateMembers(
+					{ createContainerRuntimeVersion: "1.2.3", sweepEnabled: true },
+					{ gcEnforcementMinCreateContainerRuntimeVersion: "2.0.0" },
```
Comment (agarwal-navin) [inline packages/runtime/container-runtime/src/test/garbageCollection.spec.ts:237]: gcEnforcementMinCreateContainerRuntimeVersion is too wordy IMO :)

**Review by agarwal-navin — APPROVED** [review]

File: packages/runtime/container-runtime/src/test/garbageCollection.spec.ts (hunk 3 above)
Comment (tyler-cai-microsoft) [inline packages/runtime/container-runtime/src/test/garbageCollection.spec.ts:237]: So the biggest problem with using string values here is either you have to write your own semver comparison algorithm, which is going to be challenging, or you're going to have to import a semver comparison algorithm which will cause the bundle size to increase.

Comment (markfields) [reply packages/runtime/container-runtime/src/test/garbageCollection.spec.ts:237]: Open to suggestions.

I like it because it's ugly.  It's undocumented so I went for highest degree of clarity with what it does (without writing a complete sentence)

File: packages/test/test-end-to-end-tests/src/test/gc/gcTombstoneDataStores.spec.ts (hunk 2 above)
Comment (markfields) [reply packages/test/test-end-to-end-tests/src/test/gc/gcTombstoneDataStores.spec.ts:658]: Yes.  But we'll want to log something about this option and the persisted value so we can filter it out.

File: packages/runtime/container-runtime/src/test/garbageCollection.spec.ts [hunk 4]
```diff
@@ -230,6 +231,44 @@ describe("Garbage Collection Tests", () => {
 				assert(gc.gcEnabled, "gcEnabled incorrect");
 				assert.equal(gc.latestSummaryGCVersion, 1, "latestSummaryGCVersion incorrect");
 			});
+			describe("gcEnforcementMinCreateContainerRuntimeVersion overriding sweepEnabled true for some existing containers", () => {
+				it("gcEnforcementMinCreateContainerRuntimeVersion later than persisted value: sweepEnabled false", () => {
```
Comment (markfields) [inline packages/runtime/container-runtime/src/test/garbageCollection.spec.ts:235]: This illustrates how GC enforcement would be disabled for an old document.  It's using the `sweepEnabled` setting but would also apply to tombstone failure.

File: packages/runtime/container-runtime/src/test/garbageCollection.spec.ts [hunk 5]
```diff
@@ -230,6 +231,44 @@ describe("Garbage Collection Tests", () => {
 				assert(gc.gcEnabled, "gcEnabled incorrect");
 				assert.equal(gc.latestSummaryGCVersion, 1, "latestSummaryGCVersion incorrect");
 			});
+			describe("gcEnforcementMinCreateContainerRuntimeVersion overriding sweepEnabled true for some existing containers", () => {
```
Comment (markfields) [inline packages/runtime/container-runtime/src/test/garbageCollection.spec.ts:234]: https://github.com/microsoft/FluidFramework/pull/13985/files/9fd0e174f7105671eacb0f1fe86efd5cee9bc283#r1096322059

Highlighting Navin's comment about the name being wordy - I like it but open to other suggestions :)

File: packages/runtime/container-runtime/src/test/garbageCollection.spec.ts [hunk 6]
```diff
@@ -230,6 +231,44 @@ describe("Garbage Collection Tests", () => {
...
+				it("gcEnforcementMinCreateContainerRuntimeVersion equal to persisted value: sweepEnabled true", () => {
+					gc = createGcWithPrivateMembers(
+						{ createContainerRuntimeVersion: "1.2.3", sweepEnabled: true },
+						{ gcEnforcementMinCreateContainerRuntimeVersion: "1.2.3" },
+					);
+					assert(gc.sweepEnabled, "sweepEnabled incorrect");
+				});
+				it("gcEnforcementMinCreateContainerRuntimeVersion earlier than persisted value: sweepEnabled true", () => {
+					gc = createGcWithPrivateMembers(
+						{ createContainerRuntimeVersion: "1.2.3", sweepEnabled: true },
```
Comment (markfields) [inline packages/runtime/container-runtime/src/test/garbageCollection.spec.ts:252]: Tyler [says](https://github.com/microsoft/FluidFramework/pull/13985/files/9fd0e174f7105671eacb0f1fe86efd5cee9bc283#r1096322964):

> So the biggest problem with using string values here is either you have to write your own semver comparison algorithm, which is going to be challenging as we are doing 2.0.0-internal.x.y.z and you will need to differentiate between dev versions and the like, or you're going to have to import a semver comparison algorithm which will cause the bundle size to increase.

**Comment by markfields** [comment]
Did #14050 instead
